### ISSUE TITLE
Signature of the body and declaration in a method implementation do not match.

### ISSUE THREAD

**Issue by fubar-coder** [GeirGrusom/PlatformInvoker#3]
Here's the full stack trace:

```
System.TypeInitializationException: The type initializer for 'PamTests.Npam.PamSession' threw an exception. ---> System.TypeLoadException: Signature of the body and declaration in a method implementation do not match.  Type: 'IPam_Implementation'.  Assembly: 'DynamicInterfaces, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at System.Reflection.Emit.TypeBuilder.TermCreateClass(RuntimeModule module, Int32 tk, ObjectHandleOnStack type)
   at System.Reflection.Emit.TypeBuilder.CreateTypeNoLock()
   at System.Reflection.Emit.TypeBuilder.CreateTypeInfo()
   at Platform.Invoke.LibraryInterfaceMapper.Implement[TInterface](ILibrary library, Object[] additionalConstructorArguments)
   at PamTests.Npam.Interop.Pam.Create() in /mnt/d/Test-Projekte/PamTests/PamTests/Npam/Interop/Pam.cs:line 29
   at PamTests.Npam.PamSession..cctor() in /mnt/d/Test-Projekte/PamTests/PamTests/Npam/PamSession.cs:line 16
   --- End of inner exception stack trace ---
   at PamTests.Npam.PamSession.Start(String user) in /mnt/d/Test-Projekte/PamTests/PamTests/Npam/PamSession.cs:line 49
   at PamTests.Program.TestPam(IServiceProvider serviceProvider) in /mnt/d/Test-Projekte/PamTests/PamTests/Program.cs:line 132
   at PamTests.Program.TestPam(Boolean useTest) in /mnt/d/Test-Projekte/PamTests/PamTests/Program.cs:line 125
   at PamTests.Program.Main(String[] args) in /mnt/d/Test-Projekte/PamTests/PamTests/Program.cs:line 28
```

The reason for this error was having a function with the following sigature:

```csharp
PamStatus pam_start(
    [NotNull] [MarshalAs(UnmanagedType.LPUTF8Str)] string serviceName,
    [MarshalAs(UnmanagedType.LPUTF8Str)] string user,
    in PamConv conversation,
    out IntPtr pamHandle);
```

with `PamConv` being a struct. The `in structType` parameter style is [new in C# 7.2](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7-2#reference-semantics-with-value-types).

**Comment by GeirGrusom** [comment]
Minimal reproduction:

`void foo(in int i)`

It's definately the `in` keyword that causes the issue in this case. What exactly the difference between `in` and `ref` or `out` is that causes the issue is unclear, but I'm investigating.

**Comment by GeirGrusom** [comment]
I thought this would simply be something I forgot, or missed, but it's actually a little bit deeper.

This is basically what this library would end up doing:

```asm
.method instance void Foo ([in] int32& 'value') cil managed
```

However this is what C# 7.2 does:

```asm
.method instance void Foo (int32& modreq([mscorlib]System.Runtime.InteropServices.InAttribute) 'value') cil managed
```

The obvious difference is `modreq` which basically says that the attribute is required. However the important part is expressed in the [ECMA standard](http://jilc.sourceforge.net/ecma_p2_cil.shtml#_Toc524940460):

> Two signatures that differ only by the addition of a custom modifier (required or optional) shall not be considered to match.

So this is the reason why it fails, and why it works with `ref` and `out`: they don't use `modreq`.

Now the issue I'm trying to solve is: how do I emit `modreq` attributes with System.Reflection.Emit? There is no obvious way to do it.

**Comment by fubar-coder** [comment]
Maybe this [TypeBuilder.DefineMethod](https://msdn.microsoft.com/de-de/library/x7ea0946.aspx) can help? It has a `parameterTypeRequiredCustomModifiers` parameter which might be what you're looking for.

**Comment by GeirGrusom** [comment]
Excellent! I was looking around at ParameterBuilder and found nothing. I'll see if I can get this fixed this afternoon.

**Comment by GeirGrusom** [comment]
Fixed in version 1.2.
